PUB-1915 - README updates (#152)
hmcts/pip-channel-management: README.md

  - [Requesting a bearer token](#requesting-a-bearer-token)
  - [Using the bearer token](#using-the-bearer-token)
- [Deployment](#deployment)
- [Monitoring and Logging](#monitoring-and-logging)
- [Security & Quality Considerations](#security--quality-considerations)
- [Test Suite](#test-suite)
  - [Unit tests](#unit-tests)
  - [Functional tests](#functional-tests)
- [Contributing](#contributing)
- [License](#license)

## Overview
`pip-channel-management` is a microservice that deals with operations related to retrieving subscription channels, and the generation of alternative publishing formats used throughout the subscription process (such as PDFs).

It sits within the Court and Tribunal Hearings Service (CaTH hereafter), written with Spring Boot/Java.

In practice, the service is usually containerized within a hosted kubernetes environment within Azure.

Broadly speaking, this service has two main components detailed below:

##### Subscription Channels
- For email subscriptions, it maps the subscriptions to their related email address.
- For third party subscription, it retrieves the API address to send the subscription to.

##### Alternative publishing formats

> - The generation (and associated CRUD operations) of PDFs and CSVs for JSON publications.

Reviewer: Should be PDFs and Excel spreadsheets, not CSVs

Author: Updated

> - The generation of a publication summary for JSON publications, which is used in the subscriptions process.

Reviewer: Think it should be subscription process, not subscriptions process :)

Author: Updated :)

All interactions with `pip-channel-management` are performed through the API (specified in [API Documentation](#api-documentation)) either as a standalone service or via connections to other microservices.

## Features and Functionality

> - Generation of PDFs from JSON, for each list type. These are used in the subscriptions process, and also via the Frontend

Reviewer 2: ```suggestion
- Generation of PDFs from JSON, for each list type. These are used in the subscriptions process, and also via the Frontend
```
Commas can both be removed.

Author: Updated :)

> - Generation of the Artefact Summary, which is the text that contains details about a JSON publication. This is used in the subscription email.

Reviewer: Think the sentence can be changed slightly "Generation of publication summary, which contains details about...."

Author: Updated

> - Maps a list of subscriptions to their appropriate email. It also de-duplicates the list, which in turn means only one email is sent per user.

Author: Updated

> - Maps a third party subscription to it's appropriate API endpoint.

Reviewer 2: ```suggestion
- Maps a third party subscription to its appropriate API endpoint.
```

Author: Updated

> - Maps a third party subscription to it's appropriate API endpoint.

Reviewer 2: ```suggestion
- Mapping a third party subscription to its appropriate API endpoint.
```

Author: Updated

- Azure Blob Storage: Handles interactions with the CaTH Azure Blob Storage instance (or local Azurite emulator/Azure Storage Explorer instances)
- Secure/Insecure Mode: Use of bearer tokens for authentication with the secure instance (if desired)
- OpenAPI Spec/Swagger-UI: Documents and allows users or developers to access API resources within the browser.
- Integration tests, which talk to Data Management in staging to retrieve publications.

## JSON to PDF + Summary Conversion

> Each List Type has a related converter and summary generated. This is stored in the List Type enum [here](./src/main/java/uk/gov/hmcts/reform/pip/channel/management/models/external/datamanagement/ListType.java)

Reviewer: Just a note that this location is going to change after my ticket to move shared code to. pip-data-models

Author: Good shout, will keep that in mind when it goes through sign off

> The converter does a very similar job to the frontend style guides. It maps the JSON data to an associated accessible PDF, and also CSVs for certain list types.

Reviewer: Should be Excel spreadsheets not CSVs

Author: Updated

| Variable                  | Description                                                                            |
|:--------------------------|:---------------------------------------------------------------------------------------|
| CLIENT_ID                 | As above                                                                               |
| CLIENT_SECRET             | As above                                                                               |
| APP_URI                   | As above                                                                               |
| TENANT_ID                 | As above                                                                               |
| ACCOUNT_MANAGEMENT_AZ_API | As above                                                                               |
| DATA_MANAGEMENT_AZ_API    | As above                                                                               |
| VERIFIED_USER_ID          | Value for the provenance of a verified used as a header on authentication-bound tests. |

#### Application.yaml files
The service can also be adapted using the yaml files found in the following locations:

> - `src/main/resources/application.yaml` for changes to the behaviour of the service itself.

Reviewer 2: ```suggestion
- [src/main/resources/application.yaml](./src/main/resources/application.yaml)( for changes to the behaviour of the service itself.
```

Author: Updated

- [src/main/resources/application-dev.yaml](./src/main/resources/application-dev.yaml) for changes to the behaviour of the service when running locally.
- [src/test/resources/application-test.yaml](./src/test/resources/application-test.yaml) for changes to other test types (e.g. unit tests).

## API Documentation
Our full API specification can be found within our Swagger-UI page.

> It can be accessed locally by starting the service and going to [http://localhost:8090/swagger-ui/swagger-ui/index.html](http://localhost:8181/swagger-ui/swagger-ui/index.html)

Reviewer: URL port is wrong. Should be localhost:8181 not 8090
